### PR TITLE
CC-16985: Unzer Glue support integration

### PR DESCRIPTION
**Pull request by volodymyr-bezuhlyi**
#### Overview

- Ticket: https://spryker.atlassian.net/browse/CC-16985

###### Optional

Reference all core PRs for auto-link back-references.

Core PRs
- https://github.com/spryker/suite-nonsplit/pull/6882

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -6,6 +6,7 @@
     "require": {
         "php": ">=7.4",
         "spryker-eco/unzer": "^1.0.0",
+        "spryker/checkout-rest-api-extension": "^1.5",
```
Comment (AsonUnique) [inline composer.json]:
```suggestion
        "spryker/checkout-rest-api-extension": "^1.5.0",
```

Comment (AsonUnique) [reply composer.json]: Also, the `1.3.0` version should be enough.

**Review by AsonUnique — APPROVED** [review]

File: src/SprykerEco/Glue/UnzerRestApi/Plugin/CheckoutRestApi/UnzerCheckoutDataResponseMapperPlugin.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+use Generated\Shared\Transfer\RestCheckoutDataResponseAttributesTransfer;
+use Generated\Shared\Transfer\RestCheckoutDataTransfer;
+use Generated\Shared\Transfer\RestCheckoutRequestAttributesTransfer;
+use Spryker\Glue\CheckoutRestApiExtension\Dependency\Plugin\CheckoutDataResponseMapperPluginInterface;
+use Spryker\Glue\Kernel\AbstractPlugin;
+
+class UnzerCheckoutDataResponseMapperPlugin extends AbstractPlugin implements CheckoutDataResponseMapperPluginInterface
+{
+    /**
+     * Specification:
```
Comment (ilyakubanov) [inline src/SprykerEco/Glue/UnzerRestApi/Plugin/CheckoutRestApi/UnzerCheckoutDataResponseMapperPlugin.php]:
```suggestion
     * {@inheritDoc}
```

File: src/SprykerEco/Glue/UnzerRestApi/Plugin/CheckoutRestApi/UnzerCheckoutDataResponseMapperPlugin.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        RestCheckoutDataTransfer $restCheckoutDataTransfer,
+        RestCheckoutRequestAttributesTransfer $restCheckoutRequestAttributesTransfer,
+        RestCheckoutDataResponseAttributesTransfer $restCheckoutResponseAttributesTransfer
+    ): RestCheckoutDataResponseAttributesTransfer {
+        $unzerPublicKey = $restCheckoutDataTransfer->getQuoteOrFail()
+            ->getUnzerCredentialsOrFail()
+            ->getUnzerKeypairOrFail()
+            ->getPublicKeyOrFail();
+
+        return $restCheckoutResponseAttributesTransfer->setUnzerPublicKey($unzerPublicKey);
```
Comment (ilyakubanov) [inline src/SprykerEco/Glue/UnzerRestApi/Plugin/CheckoutRestApi/UnzerCheckoutDataResponseMapperPlugin.php]: please create a mapper under `Processor\Mapper` folder and move the implementation to the mapper

File: src/SprykerEco/Glue/UnzerRestApi/Plugin/CheckoutRestApi/UnzerCheckoutDataResponseMapperPlugin.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+use Generated\Shared\Transfer\RestCheckoutDataTransfer;
+use Generated\Shared\Transfer\RestCheckoutRequestAttributesTransfer;
+use Spryker\Glue\CheckoutRestApiExtension\Dependency\Plugin\CheckoutDataResponseMapperPluginInterface;
+use Spryker\Glue\Kernel\AbstractPlugin;
+
+class UnzerCheckoutDataResponseMapperPlugin extends AbstractPlugin implements CheckoutDataResponseMapperPluginInterface
+{
+    /**
+     * Specification:
+     * - Fills `RestCheckoutDataResponseAttributesTransfer.UnzerPublicKey` property using `UnzerCredentials`, `UnzerKeypair`, `PublicKey` properties of `RestCheckoutDataResponseTransfer`.
```
Comment (ilyakubanov) [inline src/SprykerEco/Glue/UnzerRestApi/Plugin/CheckoutRestApi/UnzerCheckoutDataResponseMapperPlugin.php]:
```suggestion
     * - Maps `RestCheckoutDataTransfer.unzerCredentials.unzerKeypair.publicKey` to  `RestCheckoutDataResponseAttributesTransfer.unzerPublicKey`.
```

File: src/SprykerEco/Glue/UnzerRestApi/Plugin/CheckoutRestApi/UnzerCheckoutDataResponseMapperPlugin.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+     * @param \Generated\Shared\Transfer\RestCheckoutDataTransfer $restCheckoutDataTransfer
+     * @param \Generated\Shared\Transfer\RestCheckoutRequestAttributesTransfer $restCheckoutRequestAttributesTransfer
+     * @param \Generated\Shared\Transfer\RestCheckoutDataResponseAttributesTransfer $restCheckoutResponseAttributesTransfer
+     *
+     * @return \Generated\Shared\Transfer\RestCheckoutDataResponseAttributesTransfer
+     */
+    public function mapRestCheckoutDataResponseTransferToRestCheckoutDataResponseAttributesTransfer(
+        RestCheckoutDataTransfer $restCheckoutDataTransfer,
+        RestCheckoutRequestAttributesTransfer $restCheckoutRequestAttributesTransfer,
+        RestCheckoutDataResponseAttributesTransfer $restCheckoutResponseAttributesTransfer
```
Comment (ilyakubanov) [inline src/SprykerEco/Glue/UnzerRestApi/Plugin/CheckoutRestApi/UnzerCheckoutDataResponseMapperPlugin.php]:
```suggestion
        RestCheckoutDataResponseAttributesTransfer $restCheckoutDataResponseAttributesTransfer
```

**Review by ilyakubanov — COMMENTED** [review]
- please create a dev epic branch and change the target

File: src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/RestCheckoutDataResponseAttributesMapper.php [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ * MIT License
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace SprykerEco\Glue\UnzerRestApi\Processor\Mapper;
+
+use Generated\Shared\Transfer\RestCheckoutDataResponseAttributesTransfer;
+use Generated\Shared\Transfer\RestCheckoutDataTransfer;
+
+class RestCheckoutDataResponseAttributesMapper implements RestCheckoutDataResponseAttributesMapperInterface
```
Comment (ilyakubanov) [inline src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/RestCheckoutDataResponseAttributesMapper.php:13]:
```suggestion
class CheckoutDataResponseMapper implements CheckoutDataResponseMapperInterface
```
Don't forget to update factory method name

File: src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/RestCheckoutDataResponseAttributesMapper.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+{
+    /**
+     * @param \SprykerEco\Glue\UnzerRestApi\Processor\Mapper\RestCheckoutDataResponseAttributesTransfer $restCheckoutDataResponseAttributesTransfer
+     * @param \SprykerEco\Glue\UnzerRestApi\Processor\Mapper\RestCheckoutDataTransfer $restCheckoutDataTransfer
+     *
+     * @return \SprykerEco\Glue\UnzerRestApi\Processor\Mapper\RestCheckoutDataResponseAttributesTransfer
+     */
+    public function mapRestCheckoutData(
```
Comment (ilyakubanov) [inline src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/RestCheckoutDataResponseAttributesMapper.php]:
```suggestion
    public function mapRestCheckoutDataTransferToRestCheckoutDataResponseAttributesTransfer(
```

**Review by ilyakubanov — APPROVED** [review]

File: src/SprykerEco/Glue/UnzerRestApi/Plugin/CheckoutRestApi/UnzerCheckoutDataResponseMapperPlugin.php [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * @method \SprykerEco\Glue\UnzerRestApi\UnzerRestApiFactory getFactory()
+ */
+class UnzerCheckoutDataResponseMapperPlugin extends AbstractPlugin implements CheckoutDataResponseMapperPluginInterface
+{
+    /**
+     * {@inheritDoc}
+     * - Maps `RestCheckoutDataTransfer.unzerCredentials.unzerKeypair.publicKey` to `RestCheckoutDataResponseAttributesTransfer.unzerPublicKey`.
+     *
```
Comment (kkicheglovspryker) [inline src/SprykerEco/Glue/UnzerRestApi/Plugin/CheckoutRestApi/UnzerCheckoutDataResponseMapperPlugin.php:25]: Please add required fields which should be set for method parameters

File: src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/CheckoutDataResponseMapper.php [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+class CheckoutDataResponseMapper implements RestCheckoutDataResponseAttributesMapperInterface
+{
+    /**
+     * @param \SprykerEco\Glue\UnzerRestApi\Processor\Mapper\RestCheckoutDataResponseAttributesTransfer $restCheckoutDataResponseAttributesTransfer
+     * @param \SprykerEco\Glue\UnzerRestApi\Processor\Mapper\RestCheckoutDataTransfer $restCheckoutDataTransfer
+     *
+     * @return \SprykerEco\Glue\UnzerRestApi\Processor\Mapper\RestCheckoutDataResponseAttributesTransfer
+     */
+    public function mapRestCheckoutDataTransferToRestCheckoutDataResponseAttributesTransfer(
+        RestCheckoutDataResponseAttributesTransfer $restCheckoutDataResponseAttributesTransfer,
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/CheckoutDataResponseMapper.php]: method arguments order must be exchanged
map {1st param} to {2nd param}

File: src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/CheckoutDataResponseMapper.php [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        RestCheckoutDataTransfer $restCheckoutDataTransfer
+    ): RestCheckoutDataResponseAttributesTransfer {
+        $unzerPublicKey = $restCheckoutDataTransfer->getQuoteOrFail()
+            ->getUnzerCredentialsOrFail()
+            ->getUnzerKeypairOrFail()
+            ->getPublicKeyOrFail();
+        $restCheckoutDataResponseAttributesTransfer->setUnzerPublicKey($unzerPublicKey);
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/CheckoutDataResponseMapper.php]:
```suggestion
        
        return $restCheckoutDataResponseAttributesTransfer->setUnzerPublicKey($unzerPublicKey);
```

please do not forget empty line before return

File: src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/CheckoutDataResponseMapper.php [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+ * MIT License
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace SprykerEco\Glue\UnzerRestApi\Processor\Mapper;
+
+use Generated\Shared\Transfer\RestCheckoutDataResponseAttributesTransfer;
+use Generated\Shared\Transfer\RestCheckoutDataTransfer;
+
+class CheckoutDataResponseMapper implements RestCheckoutDataResponseAttributesMapperInterface
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/CheckoutDataResponseMapper.php]:
```suggestion
class RestCheckoutDataResponseAttributesMapper implements RestCheckoutDataResponseAttributesMapperInterface
```

Comment (DmitryLymarenko) [reply src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/CheckoutDataResponseMapper.php]: please do not forget to adjust respective factory method

File: src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/RestCheckoutDataResponseAttributesMapper.php (hunk 6 above)
Comment (michbeck) [reply src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/RestCheckoutDataResponseAttributesMapper.php:13]: Resolved in consultation with Ilya Kubanov.

File: src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/RestCheckoutDataResponseAttributesMapper.php (hunk 7 above)
Comment (michbeck) [reply src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/RestCheckoutDataResponseAttributesMapper.php]: Resolved in consultation with Ilya Kubanov.

File: src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/CheckoutDataResponseMapper.php (hunk 11 above)
Comment (michbeck) [reply src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/CheckoutDataResponseMapper.php]: Fixed.

File: src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/CheckoutDataResponseMapper.php (hunk 10 above)
Comment (michbeck) [reply src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/CheckoutDataResponseMapper.php]: Fixed + additionally wrong docblock.

File: src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/CheckoutDataResponseMapper.php (hunk 9 above)
Comment (michbeck) [reply src/SprykerEco/Glue/UnzerRestApi/Processor/Mapper/CheckoutDataResponseMapper.php]: Fixed.

File: src/SprykerEco/Glue/UnzerRestApi/Plugin/CheckoutRestApi/UnzerCheckoutDataResponseMapperPlugin.php (hunk 8 above)
Comment (michbeck) [reply src/SprykerEco/Glue/UnzerRestApi/Plugin/CheckoutRestApi/UnzerCheckoutDataResponseMapperPlugin.php:25]: Fixed.

File: src/SprykerEco/Glue/UnzerRestApi/Plugin/CheckoutRestApi/UnzerCheckoutDataResponseMapperPlugin.php [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * @method \SprykerEco\Glue\UnzerRestApi\UnzerRestApiFactory getFactory()
+ */
+class UnzerCheckoutDataResponseMapperPlugin extends AbstractPlugin implements CheckoutDataResponseMapperPluginInterface
+{
+    /**
+     * {@inheritDoc}
+     * - Maps `RestCheckoutDataTransfer.unzerCredentials.unzerKeypair.publicKey` to `RestCheckoutDataResponseAttributesTransfer.unzerPublicKey`.
+     * - Requires `RestCheckoutDataTransfer.quote.unzerCredentials.unzerKeypair.publicKey` to be set.
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Glue/UnzerRestApi/Plugin/CheckoutRestApi/UnzerCheckoutDataResponseMapperPlugin.php:23]: please make this line 1st at spec

Comment (michbeck) [reply src/SprykerEco/Glue/UnzerRestApi/Plugin/CheckoutRestApi/UnzerCheckoutDataResponseMapperPlugin.php:23]: Fixed.

**Review by DmitryLymarenko — APPROVED** [review]

**Pull request merged**